### PR TITLE
Add "Amazon Corretto" variants

### LINKED ISSUE #162: Amazon Corretto variant support
Is this possible? 

https://hub.docker.com/_/amazoncorretto
https://github.com/corretto/corretto-8-docker

### PR DESCRIPTION
**Pull request by tianon**
Closes #162

cc @jguo11 @ziyiluo @alvdavi (`amazoncorretto` image maintainers) -- if any of you are willing/able to provide any thoughts or input on this, it would certainly be appreciated! :+1:

also cc @iliana @praveen-pk @Eric-Warehime (`amazonlinux` image maintainers) in case you care to provide any additional thoughts :innocent: :heart:

(There's still a couple `TODO` items here that I'm not 100% satisfied with. :sweat_smile:)

### CODE REVIEW

**Comment by tianon** [comment]
The idea is that this will lead to image tags like `tomcat:7-jdk8-corretto`, etc.

**Comment by tianon** [comment]
<details>
<summary>A `Dockerfile` diff that might be easier to review:</summary>

```diff
$ diff -u Dockerfile-apt.template Dockerfile-yum.template
--- Dockerfile-apt.template	2019-06-24 17:13:39.048783141 -0700
+++ Dockerfile-yum.template	2019-06-24 17:13:39.048783141 -0700
@@ -34,21 +34,14 @@
 
 RUN set -eux; \
 	\
-	savedAptMark="$(apt-mark showmanual)"; \
-	apt-get update; \
-	\
-	apt-get install -y --no-install-recommends gnupg dirmngr; \
-	\
 	export GNUPGHOME="$(mktemp -d)"; \
 	for key in $GPG_KEYS; do \
 		gpg --batch --keyserver ha.pool.sks-keyservers.net --recv-keys "$key"; \
 	done; \
 	\
-	apt-get install -y --no-install-recommends wget ca-certificates; \
-	\
 	success=; \
 	for url in $TOMCAT_TGZ_URLS; do \
-		if wget -O tomcat.tar.gz "$url"; then \
+		if curl -fL -o tomcat.tar.gz "$url"; then \
 			success=1; \
 			break; \
 		fi; \
@@ -59,13 +52,16 @@
 	\
 	success=; \
 	for url in $TOMCAT_ASC_URLS; do \
-		if wget -O tomcat.tar.gz.asc "$url"; then \
+		if curl -fL -o tomcat.tar.gz.asc "$url"; then \
 			success=1; \
 			break; \
 		fi; \
 	done; \
 	[ -n "$success" ]; \
 	\
+# TODO should we leave these installed, or remove them too?
+	yum install -y gzip tar; \
+	\
 	gpg --batch --verify tomcat.tar.gz.asc tomcat.tar.gz; \
 	tar -xvf tomcat.tar.gz --strip-components=1; \
 	rm bin/*.bat; \
@@ -75,20 +71,18 @@
 	\
 	nativeBuildDir="$(mktemp -d)"; \
 	tar -xvf bin/tomcat-native.tar.gz -C "$nativeBuildDir" --strip-components=1; \
-	apt-get install -y --no-install-recommends \
-		dpkg-dev \
+	nativeBuildDeps=' \
+		apr-devel \
 		gcc \
-		libapr1-dev \
-		libssl-dev \
 		make \
-	; \
+		openssl-devel \
+	'; \
+	yum install -y $nativeBuildDeps; \
 	( \
 		export CATALINA_HOME="$PWD"; \
 		cd "$nativeBuildDir/native"; \
-		gnuArch="$(dpkg-architecture --query DEB_BUILD_GNU_TYPE)"; \
-		aprConfig="$(which apr-1-config)"; \
+		aprConfig="$(type -p apr-1-config)"; \
 		./configure \
-			--build="$gnuArch" \
 			--libdir="$TOMCAT_NATIVE_LIBDIR" \
 			--prefix="$CATALINA_HOME" \
 			--with-apr="$aprConfig" \
@@ -100,19 +94,19 @@
 	rm -rf "$nativeBuildDir"; \
 	rm bin/tomcat-native.tar.gz; \
 	\
-# reset apt-mark's "manual" list so that "purge --auto-remove" will remove all build dependencies
-	apt-mark auto '.*' > /dev/null; \
-	[ -z "$savedAptMark" ] || apt-mark manual $savedAptMark > /dev/null; \
 	find "$TOMCAT_NATIVE_LIBDIR" -type f -executable -exec ldd '{}' ';' \
 		| awk '/=>/ { print $(NF-1) }' \
 		| sort -u \
-		| xargs -r dpkg-query --search \
-		| cut -d: -f1 \
+		| xargs -r rpm -q --queryformat '%{NAME}-%{VERSION}-%{RELEASE}-%{ARCH}\n' --whatprovides \
 		| sort -u \
-		| xargs -r apt-mark manual \
+# http://yum.baseurl.org/wiki/YumDB.html (updating "reason" to "user" for tomcat-native build dependencies)
+# TODO find a better/cleaner way to generically interact with the yumdb for the purposes of marking things as automatically/manually installed
+		| xargs -r -n1 bash -Eeuo pipefail -c 'yumdbDir=( "/var/lib/yum/yumdb/${1:0:1}/"*"-$1" ); if [ -d "$yumdbDir" ]; then echo "note: updating $yumdbDir/reason"; echo -n user > "$yumdbDir/reason"; else echo >&2 "warning: $yumdbDir does not exist (ignoring)"; fi' -- \
 	; \
-	apt-get purge -y --auto-remove -o APT::AutoRemove::RecommendsImportant=false; \
-	rm -rf /var/lib/apt/lists/*; \
+# TODO handle "$nativeBuildDeps" in a similarly generic way?
+	yum autoremove -y $nativeBuildDeps; \
+	yum clean all; \
+	rm -rf /var/cache/yum; \
 	\
 # sh removes env vars it doesn't support (ones with periods)
 # https://github.com/docker-library/tomcat/issues/77
````

</details>

File: 8.5/jdk11/corretto/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+	success=; \
+	for url in $TOMCAT_ASC_URLS; do \
+		if curl -fL -o tomcat.tar.gz.asc "$url"; then \
+			success=1; \
+			break; \
+		fi; \
+	done; \
+	[ -n "$success" ]; \
+	\
+# TODO should we leave these installed, or remove them too?
```
Comment (jguo11) [inline 8.5/jdk11/corretto/Dockerfile]: I don't see you remove them in the dockerfile of other distributions.

**Comment by tianon** [comment]
a2fbaca takes us from ~590MB down to ~521MB

(https://travis-ci.org/docker-library/tomcat/jobs/550008647#L2270 vs https://travis-ci.org/docker-library/tomcat/jobs/550702570#L1492)

For comparison:

- `tomcat:9.0-jdk11-openjdk`: ~623MB (https://travis-ci.org/docker-library/tomcat/jobs/550008643#L2562)
- `tomcat:9.0-jdk11-openjdk-slim`: ~404MB (https://travis-ci.org/docker-library/tomcat/jobs/550008644#L2762)
- `tomcat:9.0-jdk11-adoptopenjdk-hotspot`: ~442MB (https://travis-ci.org/docker-library/tomcat/jobs/550008645#L2956)

**Pull request merged**